instrumentation/opensearch/opensearch-java-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opensearch/v3_0/OpenSearchBodyExtractor.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.opensearch.v3_0;
+
+import static java.util.logging.Level.FINE;
+
+import jakarta.json.stream.JsonGenerator;
+import java.io.ByteArrayOutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import org.opensearch.client.json.JsonpMapper;
+import org.opensearch.client.json.NdJsonpSerializable;
+import org.opensearch.client.transport.GenericSerializable;
+
+public final class OpenSearchBodyExtractor {
+
+  private static final Logger logger = Logger.getLogger(OpenSearchBodyExtractor.class.getName());
+
+  @Nullable
+  public static String extract(JsonpMapper mapper, Object request) {
+    try {
+      ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+      if (request instanceof NdJsonpSerializable) {
+        writeNdJson(mapper, (NdJsonpSerializable) request, baos);
+      } else if (request instanceof GenericSerializable) {
+        ((GenericSerializable) request).serialize(baos);
+      } else {
+        JsonGenerator generator = mapper.jsonProvider().createGenerator(baos);
+        mapper.serialize(request, generator);
+        generator.close();
+      }
+
+      String body = baos.toString(StandardCharsets.UTF_8);
+      return body.isEmpty() ? null : body;
+    } catch (RuntimeException e) {
+      logger.log(FINE, "Failure extracting body", e);
+      return null;
+    }
+  }
+
+  private static void writeNdJson(
+      JsonpMapper mapper, NdJsonpSerializable value, ByteArrayOutputStream baos) {
+    try {
+      Iterator<?> values = value._serializables();
+      while (values.hasNext()) {
+        Object item = values.next();
+        if (item instanceof NdJsonpSerializable && item != value) {
+          // do not recurse on the item itself
+          writeNdJson(mapper, (NdJsonpSerializable) item, baos);
+        } else {
+          JsonGenerator generator = mapper.jsonProvider().createGenerator(baos);
```

Comment:
Just wondering whether instead of serializing the query to json string and then parsing that json, sanitizing it and serializing it again to json string you could wrap the `JsonGenerator` here so that it would replace literal values with `?` so you wouldn't need to parse the json again.